pytorch_lightning/experimental/plugins/secure/pysyft.py

```diff
@@ -0,0 +1,121 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, Generator
+
+import torch
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from pytorch_lightning.experimental.plugins.secure.base import BaseSecurePlugin
+from pytorch_lightning.utilities.imports import _PYSYFT_AVAILABLE
+
+
+class PySyftPlugin(BaseSecurePlugin):
+
+    @staticmethod
+    def optimizer_state(optimizer: Optimizer) -> Dict[str, torch.Tensor]:
+        """
+        Returns state of an optimizer. Allows for syncing/collating optimizer state from processes in custom
+        plugins.
+        """
+        return optimizer.state_dict().get(request_block=True, delete_obj=False)
+
+    @staticmethod
+    def save_function(trainer, filepath: str, save_weights_only: bool) -> None:
+        model_ref = trainer.lightning_module
+        sy_model = model_ref.get_model()
+        sy_model.device = model_ref.device
+        sy_model.hparams = model_ref.hparams
+        sy_model.on_save_checkpoint = model_ref.on_save_checkpoint
+        trainer.training_type_plugin.model = sy_model
+        trainer.accelerator.optimizer_state = PySyftPlugin.optimizer_state
+        trainer.save_checkpoint(filepath, save_weights_only)
+        trainer.training_type_plugin.model = model_ref
+
+
+if _PYSYFT_AVAILABLE:
+
+    from types import ModuleType
+    from typing import Any, Union
+
+    import syft as sy
+
+    from pytorch_lightning.core.lightning import LightningModule
+
+    SyModuleProxyType = Union[ModuleType, Module]
+    SyModelProxyType = Union[Module, sy.Module]
+
+    # cant use lib_ast during test search time
+    TorchTensorPointerType = Any  # sy.lib_ast.torch.Tensor.pointer_type
+    SyTensorProxyType = Union[torch.Tensor, TorchTensorPointerType]  # type: ignore
+
+    class SyLightningModule(LightningModule):
+
+        def __init__(self, download_back: bool = False, run_locally: bool = False) -> None:
+            super().__init__()
+            # Those are helpers to easily work with `sy.Module`
+            self.duet = sy.client_cache["duet"]
+            self.remote_torch = sy.client_cache["duet"].torch
+            self.local_torch = globals()["torch"]
+            self.download_back = download_back
```

Comment:
what is download_back for?

Reply:
Renamed `request_parameters`. It also to pull the trained weights from the remote machine.